release.sh

```diff
@@ -0,0 +1,351 @@
+
+    REMOTE_BRANCH="origin/$TRAIN_BRANCH"
+    REMOTE_BRANCH_EXISTS=`git branch -r | awk '{$1=$1};1' | grep "^$REMOTE_BRANCH\$"` || true
+
+    if [ "$REMOTE_BRANCH_EXISTS" = "" ]; then
+      git checkout -b "$TRAIN_BRANCH"
+    else
+      git checkout --track -b "$TRAIN_BRANCH" "$REMOTE_BRANCH"
+    fi
+  else
+    git checkout "$TRAIN_BRANCH"
+    git pull origin "$TRAIN_BRANCH" || true
+  fi
+fi
+
+# 6. Check we're on a train branch.
+TRAIN_BRANCH=`git branch | grep '^\*' | cut -d ' ' -f 2`
+IS_TRAIN_BRANCH=`echo "$TRAIN_BRANCH" | cut -d '-' -f 1`
+if [ "$IS_TRAIN_BRANCH" != "train" ]; then
+  abort "This is not a train branch"
+fi
+
+LAST_VERSION="$MAJOR.$TRAIN.$PATCH"
+SED_FRIENDLY_LAST_VERSION="$MAJOR\\.$TRAIN\\.$PATCH"
+
+# 7. Generate the bumped version string.
+case "$BUILD_TYPE" in
+  "Train")
+    TRAIN=`expr $TRAIN + 1`
+    PATCH=0
+    ;;
+  "Patch")
+    PATCH=`expr $PATCH + 1`
+    ;;
+esac
+NEW_VERSION="$MAJOR.$TRAIN.$PATCH"
+NEW_TAG="v$NEW_VERSION"
+
+# 8. For each of the "main" packages...
+bump() {
+  # 8.1. List commits since the last tag.
+  LOCAL_COMMITS=`git log $LAST_TAG..HEAD --pretty=oneline --abbrev-commit -- "$1"`
+
+  # 8.2. For each commit...
+  while read -r COMMIT; do
+    HASH=`echo "$COMMIT" | cut -d ' ' -f 1`
+    MESSAGE=`echo "$COMMIT" | cut -d ':' -f 2- | awk '{$1=$1};1'`
+    TYPE=`echo "$COMMIT" | cut -d ' ' -f 2 | awk '{$1=$1};1' | cut -d ':' -f 1 | cut -d '(' -f 1 | awk '{$1=$1};1'`
+    AREA=`echo "$COMMIT" | cut -d '(' -f 2 | cut -d ')' -f 1 | awk '{$1=$1};1'`
+
+    if [ "$AREA" = "$COMMIT" ]; then
+      AREA=""
+    fi
+
+    if [ "$AREA" != "" ]; then
+      AREA="$AREA: "
+    fi
+
+    # 8.2.1. Add the commit message to a summary string.
+    case "$TYPE" in
+      "")
+        # Ignore blank lines
+        ;;
+      "Merge")
+        # Ignore merge commits
+        ;;
+      "Release")
+        # Ignore release commits
+        ;;
+      "feat")
+        if [ "$FEAT_SUMMARY" = "" ]; then
+          FEAT_SUMMARY="### New features\n"
+        fi
+        FEAT_SUMMARY="$FEAT_SUMMARY\n* $AREA$MESSAGE ($HASH)"
+        ;;
+      "fix")
+        if [ "$FIX_SUMMARY" = "" ]; then
+          FIX_SUMMARY="### Bug fixes\n"
+        fi
+        FIX_SUMMARY="$FIX_SUMMARY\n* $AREA$MESSAGE ($HASH)"
+        ;;
+      "perf")
+        if [ "$PERF_SUMMARY" = "" ]; then
+          PERF_SUMMARY="### Performance improvements\n"
+        fi
+        PERF_SUMMARY="$PERF_SUMMARY\n* $AREA$MESSAGE ($HASH)"
+        ;;
+      "refactor")
+        if [ "$REFACTOR_SUMMARY" = "" ]; then
+          REFACTOR_SUMMARY="### Refactorings\n"
+        fi
+        REFACTOR_SUMMARY="$REFACTOR_SUMMARY\n* $AREA$MESSAGE ($HASH)"
+        ;;
+      "revert")
+        if [ "$REFACTOR_SUMMARY" = "" ]; then
+          REVERT_SUMMARY="### Reverted changes\n"
+        fi
+        REVERT_SUMMARY="$REVERT_SUMMARY\n* $AREA$MESSAGE ($HASH)"
+        ;;
+      *)
+        if [ "$OTHER_SUMMARY" = "" ]; then
```

Comment:
I considered this list from the perspective of a reader of the changelog. What value is there for them in calling out "chore" as something separate to "other changes"? I didn't see any but if I'm missing something I'm happy to add it.

Reply:
I think this is fine as is, it might only really be an issue if the "other changes" section got too big that a reader would not easily see important changes.